src/amp-story-player.js

```diff
@@ -248,15 +377,112 @@ export class AmpStoryPlayer {
    * @param {number} iframeIdx
    */
   displayStory_(iframeIdx) {
-    this.messagingFor_[iframeIdx].sendRequest(
-      'visibilitychange',
-      {state: 'visible'},
-      true
-    );
+    this.handshakePromises_[iframeIdx].then(() => {
+      this.messagingFor_[iframeIdx].sendRequest(
+        'visibilitychange',
+        {state: 'visible'},
+        true
+      );
+    });
+  }
+
+  /**
+   * Sends a message to the story document to pre-render it.
+   * @private
+   * @param {number} iframeIdx
+   */
+  preRenderStory_(iframeIdx) {
+    this.handshakePromises_[iframeIdx].then(() => {
+      this.messagingFor_[iframeIdx].sendRequest(
+        'visibilitychange',
+        {state: 'prerender'},
+        true
+      );
+    });
+  }
+
+  /**
+   * Sends a message to the story document to pause it.
+   * @private
+   * @param {number} iframeIdx
+   */
+  pauseStory_(iframeIdx) {
+    this.handshakePromises_[iframeIdx].then(() => {
+      this.messagingFor_[iframeIdx].sendRequest(
+        'visibilitychange',
+        {state: 'paused'},
+        true
+      );
+    });
   }
 }
 
 self.onload = () => {
   const manager = new AmpStoryPlayerManager(self);
   manager.loadPlayers();
 };
+
+/**
+ * Manages the iframes used to host the stories inside the player. It keeps
+ * track of the iframes hosting stories, and what stories have an iframe.
+ */
+class IframePool {
```

Comment:
Let's maintain the (now controversial) one class per file rule of thumb?

Reply:
I personally don't feel strongly at all (and am actually probably on your side), but it's the repository's guidelines :man_shrugging: